### ISSUE TITLE
macOS build failure due to missing mkdtemp

### ISSUE THREAD

**Issue by falkTX** [drobilla/zix#6]
Build logs:

```
../src/posix/filesystem_posix.c:278:10: error: call to undeclared function 'mkdtemp'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
    if (!mkdtemp(result)) {
         ^
1 error generated.
```

Happens with at least macOS 14.

The command from the meson logs that was trying to build the file:
```
gcc -Ilibzix-0.a.p -I. -I.. -I../include -I/Users/runner/PawPawBuilds/targets/macos-lto/include -fvisibility=hidden -fdiagnostics-color=always -DNDEBUG -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -std=c99 -O3 -pipe -fPIC -DPIC -fdata-sections -ffunction-sections -fno-common -fvisibility=hidden -fno-stack-protector -U_FORTIFY_SOURCE -Wp,-U_FORTIFY_SOURCE -ffast-math -Os -DNDEBUG=1 -fomit-frame-pointer -fno-strict-aliasing -flto -mtune=generic -msse -msse2 -mfpmath=sse -DMAC_OS_X_VERSION_MAX_ALLOWED=MAC_OS_X_VERSION_10_8 -DMAC_OS_X_VERSION_MIN_REQUIRED=MAC_OS_X_VERSION_10_8 -mmacosx-version-min=10.8 -stdlib=libc++ -Wno-deprecated-declarations -arch x86_64 -Werror=objc-method-access -DHAVE_MIXED_SIZE_ADDRESSING -fno-finite-math-only -fPIC -pthread -DZIX_NO_DEFAULT_CONFIG -D_POSIX_C_SOURCE=200809L -D_XOPEN_SOURCE=600 -DHAVE_CLOCK_GETTIME -DHAVE_FILENO -DHAVE_LSTAT -DHAVE_MLOCK -DHAVE_PATHCONF -DHAVE_POSIX_MEMALIGN -DHAVE_REALPATH -DHAVE_SYSCONF -DZIX_STATIC -DZIX_INTERNAL -MD -MQ libzix-0.a.p/src_posix_filesystem_posix.c.o -MF libzix-0.a.p/src_posix_filesystem_posix.c.o.d -o libzix-0.a.p/src_posix_filesystem_posix.c.o -c ../src/posix/filesystem_posix.c
```

can be seen failing as part of CI here https://github.com/DISTRHO/PawPaw/actions/runs/11998315195/job/33445235997

**Comment by drobilla** [comment]
That version is old (and maybe modified?  It seems that the wrong semaphore implementation is being compiled which is wild), and zix is continuously built on MacOS: https://gitlab.com/drobilla/zix/-/jobs/8456928069

As far as I can tell, `mkdtemp` has been in MacOS since ever, but this is an especially annoying case of the `*_SOURCE` macros needing to be just right.

Please try building the latest git on this machine.

**Comment by falkTX** [comment]
that gives me a hit, this might be a meson misconfiguration on my side.

I need to provide a cross-compilation file on github CI because new runners run on ARM64 systems, but I want to have some old compat binaries for macos-intel platforms. Its one of the oddities about meson.

The way zix project checks for macos is if platform name is "darwin". I noticed a typo regarding this on my side, lets see...

**Comment by drobilla** [comment]
This is because of `-D_POSIX_C_SOURCE=200809L -D_XOPEN_SOURCE=600`

Neither zix 0.4.2 nor anything since adds that on darwin, but the log says this is 0.4.2 and doesn't show this set in CFLAGS.  I don't know how this is possible.  I double checked on my Macbook just now and neither 0.4.2 nor current `main` add that flag.

> I noticed a typo regarding this on my side, lets see...

Is this copy of zix modified in any way?

**Comment by falkTX** [comment]
> > I noticed a typo regarding this on my side, lets see...
> 
> Is this copy of zix modified in any way?

there is a small patch to forcibly hide symbols, but nothing that changes build system or real code.
the fix I am testing now is https://github.com/DISTRHO/PawPaw/commit/b6e03c7f608c1beb34a233f1e0690244ae012562 which is unrelated to zix

**Comment by drobilla** [comment]
It almost looks like you've managed to convince it to build for the wrong platform via a cross file, but that... would be quite something.

> there is a small patch to forcibly hide symbols

I'm sure it's unrelated but since I'm working on preparing a release as we speak, can you link me to this?

**Comment by falkTX** [comment]
> It almost looks like you've managed to convince it to build for the wrong platform via a cross file, but that... would be quite something.

I am surprised about that too, I really dislike how meson tries to handle cross-compilation

> I'm sure it's unrelated but since I'm working on preparing a release as we speak, can you link me to this?

Sure
https://github.com/DISTRHO/PawPaw/blob/main/patches/zix/01_hidden-zix-api.patch

Its just a simple thing to not have to `#define ZIX_API` for all builds.

**Comment by drobilla** [comment]
Good grief.  I'm sorry I asked :laughing:

**Comment by drobilla** [comment]
> the fix I am testing now is https://github.com/DISTRHO/PawPaw/commit/1138551128405c6216230fcce02638b2a70f1750 which is unrelated to zix

I see.  Apparently you can just set the system to whatever there, and meson added `macos` and other more specific, er, "subsystems" in 1.2.0.  Apparently someone realized this would be problematic, since the documentation says "Most values are meant to be used in cross files only, as those platforms can not run Meson natively".  Unfortunately this doesn't really make sense: if the values are used in cross files, then they need to be handled by build scripts (and they aren't here, hence the build failure).

tl;dr: unless you have a good specific reason to do otherwise, use `darwin`

**Comment by falkTX** [comment]
> tl;dr: unless you have a good specific reason to do otherwise, use `darwin`

this is what I understood too. but this way of specifying the build host and reading it back seems very fragile...

**Comment by drobilla** [comment]
The bunch of different values that all effectively mean "darwin" without a way of checking for that being provided is certainly fragile.  The mechanism itself seems fine to me (aside from the fact that having cross files be mandatory instead of just a way of setting options was an ugly mistake, anyway).

Anyway, I'm not about to start writing `host_machine.system() in ["darwin", "macos", "ios", "ios-simulator", "tvos", "tvos-simulator", "watchos", "watchos-simulator"]` everywhere without a good reason, so closing this.  Unfortunately it wouldn't make sense to error out here, because the catch-all `else` case will work on the vast majority of vaguely POSIX-like systems.
